<a href="https://colab.research.google.com/github/sarpit2021/SARPIT2021/blob/main/Machine_Learning_%E2%80%93_Kaggle_Titanic_Solution_Python_(Easy).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Gathering Data

In [1]:
import numpy as np
import pandas as pd

#The Machine learning alogorithm
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

# Test train split
from sklearn.model_selection import train_test_split


In [2]:
train = pd.read_csv("train.csv")
train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# 2. Data Pre-processing

In [3]:
train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Describe is a good command to get to know the data in a summarized way.


This is a case of supervised learning in which the model needs inputs and output to learn. Well in this case ‘Survived’ Column is output column and rest all are input columns.

So there are 11 input columns. However, not all columns are always important for the model to learn. If you remember the Titanic movie, you will know that the rich were more likely to survive. Also, the preference was given to children, women and aged persons.

So according to our hypothesis, older rich women and children were the most likely to survive and poor middle-aged men were the least likely to survive.

Age and Sex are directly provided in the data. We can presume whether a person is rich or poor by looking at Passenger class (Pclass).

So these are the 3 inputs to our machine learning algorithm: Passenger class, age and sex.

In [ ]:
#Checking the missing values
train.isnull().sum()

We can see that Age has 177 missing values out of 891. Thus we can do the missing values imputation. 

For now, let’s not take the Age column. So for model input, we will have only Passenger class and Sex. The output is the Survived field.


# 2.1 Data Preparation

In [4]:
# Selecting only 2 columns for ease

train_x = train[["Pclass","Sex"]]
train_x.head()

Pclass     Sex
0       3    male
1       1  female
2       3  female
3       1  female
4       3    male

Let’s extract selected 2 input columns into a new dataframe train_x.

In [6]:
# Selecting the output/ target variable
train_y = train[["Survived"]]
train_y.head()


Survived
0         0
1         1
2         1
3         1
4         0

Similarly, extract the output to train_y.

As mentioned above, the algorithms in Sklearn only work with numbers. That means we can not pass the sex as male or female.

# 2.2 Cleaning of data

In [5]:
# Making Male/ Female to integer numbers calles Label Encoding
train_x["Sex"].replace("male", 1, inplace = True)
train_x["Sex"].replace("female", 0, inplace = True)
train_x.head()

/usr/local/lib/python3.8/dist-packages/pandas/core/generic.py:6619: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


Pclass  Sex
0       3    1
1       1    0
2       3    0
3       1    0
4       3    1

We have updated the sex column.

# 2.3 Train/Test Split to prevent overfitting

We split our data into a train set and a cross-validation set. The training set is used to train the machine learning algorithm. While the cross-validation set is used to find the model accuracy (as we have the actual output for the cross-validation set). Accuracy is calculated by comparing the actual output with the predicted output. Refer this link on how accuracy is calculated in classification problem. Test set is the data for which we do not have the Output variable (Survived in this problem). You can find test csv file in the downloaded folder.

We will use the train_test_split function to create the test/ train (cross-validation) split. We will use 70% of the data to train and model and 30% of the data to check accuracy.

In [7]:
# Making dataset for validation

tr_x, cv_x, tr_y, cv_y   = train_test_split(train_x, train_y, test_size = 0.30)
print(tr_x.head())
print(tr_y.head())


     Pclass  Sex
582       2    1
6         1    1
872       1    1
87        3    1
444       3    1
     Survived
582         0
6           0
872         0
87          0
444         1


tr_x & tr_y are the training input and output and cv_x & cv_y are cross-validation input and output.

# 3.1 Random Forest Model

In [8]:
# Call the Machine Learning Algorithm
rf = RandomForestClassifier()


In [9]:
# Fitting and training the above called algorithm
rf.fit(tr_x, tr_y)

<ipython-input-9-57802e332598>:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf.fit(tr_x, tr_y)


RandomForestClassifier()

his simple fit() function is used to train our algorithm. This function takes our input dataframe (tr_x) and learns the expected output (tr_y). That’s why we narrowed the input columns so that the algorithm is not confused by the noise.

There is a popular saying in the analytics community “Garbage in Garbage out”. It describes the concept that flawed, or nonsense input data produces nonsense output or “garbage”. For avoiding this we use feature engineering and feature selection which we will cover in the next tutorial.

# 4. Evaluating the model
The instance has now “learned” how to predict Titanic survivors as the model is fitted. Now we can check how accurate our algorithm is on cross-validation data:

In [10]:
Accuracy_RandomForest = rf.score(cv_x, cv_y)
print("Accuracy = {}%".format(Accuracy_RandomForest * 100))

Accuracy = 77.23880597014924%


The score() function takes the cross-validation input and finds out the accuracy by comparing our predictive output and the known test outputs.

# Additional: Logistic Regression Model (Training & Evaluation)


This is just to show how easy it is to implement other machine learning classification models using sklearn library in python.

In [11]:
lgr = LogisticRegression()
lgr.fit(tr_x, tr_y)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression()

LogicticRegression model is fitted and we can check the accuracy on cross-validation data.

In [12]:
Accuracy_LogisticRegression = lgr.score(cv_x, cv_y)
print("Accuracy = {}%".format(Accuracy_LogisticRegression * 100))

Accuracy = 77.23880597014924%
